### PR TITLE
Link feature for imagebase

### PR DESCRIPTION
**Pull request by Comandeer**
## What is the purpose of this pull request?

New feature

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_contributing_code-section-tests),
[how to set the testing environment](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests) and
[how to create tests](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests-section-creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What changes did you make?

I've created new API for adding new features to new image widgets. Currently it's used only to add link feature to `easyimage` plugin.

The API is based on one new method in `CKEDITOR.plugins.imagebase`:

```javascript
addFeature( editor: CKEDITOR.editor, featureName: String, widgetDefinition: WidgetDefinition );
```

Every feature extends widget definition with all needed fields and callbacks. Additionally callbacks are merged together (e.g. multiple features with `init` callback will result in widget definition with `init` callback that will fire all features' init callbacks). There is also `setUp` callback, that is fire immediately and is fired once. It's suitable to e.g. add new listeners to the editor.

Link feature hooks itself to `unlink` command, however it hijacks it and stop propagation. Such solution is inspired by `image2`, however it prevents `afterCommandExec` from firing. Probably it'd be good to fire `afterCommandExec` manually.

### CODE REVIEW

File: plugins/imagebase/plugin.js [hunk 1]
```diff
@@ -104,6 +272,46 @@
 			var widget = editor.widgets.add( name, createWidgetDefinition( editor, definition ) );
 
 			editor.addFeature( widget );
+		},
+
+		/**
+		 * Adds new feature to the newly created image widget by invoking initial setup once for the editor
```
Comment (mlewand) [inline plugins/imagebase/plugin.js]: Well to be fair, you made it in a way it's compatible with any widget. So I suppose you can refer just to "widget" rather than "image widget" in this method docs.

Comment (mlewand) [reply plugins/imagebase/plugin.js]: > (...) feature to the newly created image widget by (...)

This part kinda leads me to thinking that it's feature added for a single widget instance. Assumption which is cleared later on.

But I think we should make it clear that it's for a widget definition from the very beginning.

File: plugins/imagebase/plugin.js [hunk 2]
```diff
@@ -104,6 +272,46 @@
...
+					oldOne && oldOne.apply( this, arguments );
+					newOne && newOne.apply( this, arguments );
+				};
+			}
+
+			featureDefinition.init = mergeMethods( definition.init, featureDefinition.init );
+			featureDefinition.data = mergeMethods( definition.data, featureDefinition.data );
+
+			if ( featureDefinition.setUp ) {
+				featureDefinition.setUp( editor );
```
Comment (mlewand) [inline plugins/imagebase/plugin.js]: If this function is meant to be a candidate for widget feature `setUp` function, you should also pass `definition` to it, so that widget feature has a possibility to have a look at widget that attempts to reuse it.

File: plugins/imagebase/plugin.js [hunk 3]
```diff
@@ -6,6 +6,168 @@
 ( function() {
 	'use strict';
 
+	function getFocusedWidget( editor ) {
+		return editor.widgets.focused;
+	}
+
+	function isLinkable( widget ) {
+		return widget && typeof widget.parts.link !== 'undefined';
```
Comment (mlewand) [inline plugins/imagebase/plugin.js]: I find duck typing for linking feature a bit naive. It's very possible that 3rd parties have `link` part used, in this part these will freak out.

How about providing definite a way to tell that widget implements given feature? For instance it could be simply an array in `widgetDefinition` like `widgetFeatures` with `name`s of features registered.

**Comment by mlewand** [comment]
Though there were some comments, I guess we're going in a right direction and after fixing these problems we're good to go (assuming that we won't be moving widget features to widget core just now).

File: plugins/imagebase/plugin.js [hunk 4]
```diff
@@ -6,6 +6,168 @@
...
+			linkElement.setAttributes( attributes.set );
+		}
+
+		if ( attributes.removed.length ) {
+			linkElement.removeAttributes( attributes.removed );
+		}
+	}
+
+	function createLink( editor, img, linkData ) {
+		// Covers cases when widget with link inside is upcasted.
```
[the remark this answers is not in the document]
Comment (Comandeer) [reply plugins/imagebase/plugin.js]: Actually I meant something like "covers cases when widget is created on element with link inside". I'll reword it.

File: plugins/imagebase/plugin.js [hunk 5]
```diff
@@ -6,6 +6,168 @@
...
+				a: {
+					attributes: '!href'
+				}
+			},
+			parts: {
+				link: 'a'
+			},
+
+			setUp: function( editor ) {
+				editor.on( 'dialogShow', function( evt ) {
```
[the remark this answers is not in the document]
Comment (Comandeer) [reply plugins/imagebase/plugin.js]: Such check is located in `easyimage` plugin before adding this feature.  Maybe it's sufficient?

[the remark this answers is not in the document]
Comment (Comandeer) [reply plugins/imagebase/plugin.js]: I'm wondering if we could move such checkings inside `setUp` and return boolean and then in `addFeature`:

[the remark this answers is not in the document]
Comment (mlewand) [reply plugins/imagebase/plugin.js]: > Such check is located in easyimage plugin before adding this feature. Maybe it's sufficient?

We need to ask our self if this feature is usable without `link` plugin? IMO it is usable with API, so technically only dialog hook should look at link plugin.

> I'm wondering if we could move such checkings inside setUp and return boolean and then in addFeature:

Hmm, to be honest I don't think anyone should worry about feature being applicable or not. I think that features should be "harmless" in a way that you could just carelessly apply a whole bunch of them and don't worry that something explode.

But I think it should be widget feature author responsibility to create that code in that harmless way.

tl;dr I wouldn't introduce this mechanic at this point, we don't need it (yet).

File: tests/plugins/imagebase/features/manual/link.html [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+		<a href="https://cksource.com">
+			<img src="../../../image2/_assets/foo.png" alt="foo">
+		</a>
+		<figcaption>Test image</figcaption>
+	</figure>
+</div>
+
+<script>
+	CKEDITOR.replace( 'classic', {
+		extraAllowedContent: 'figure(*)',
```
Comment (mlewand) [inline tests/plugins/imagebase/features/manual/link.html]: This line is not needed.

File: tests/plugins/imagebase/features/link.html [hunk 7]
```diff
@@ -0,0 +1,7 @@
+<div id="upcastTest">
+	<figure>
+		<a href="http://foo">
+			<img src="http://foo">
```
Comment (mlewand) [inline tests/plugins/imagebase/features/link.html]: By using this fake src your tests are causing good amount errors in dev console. It can be simply fixed by using our common assets, e.g. `%BASE_PATH%_assets/logo.png`.

**Comment by mlewand** [comment]
I have made a PR with fixes to my issues above in #1151.

File: plugins/imagebase/plugin.js [hunk 8]
```diff
@@ -104,6 +274,48 @@
...
+			featureDefinition.init = mergeMethods( definition.init, featureDefinition.init );
+			featureDefinition.data = mergeMethods( definition.data, featureDefinition.data );
+
+			if ( featureDefinition.setUp ) {
+				featureDefinition.setUp( editor, definition );
+
+				delete featureDefinition.setUp;
+			}
+
+			featureDefinition.widgetFeatures = [ name ];
```
[the remark this answers is not in the document]
Comment (Comandeer) [reply plugins/imagebase/plugin.js]: I was pretty sure that `CKEDITOR.tools.object.merge` is merging also arrays. It was quite obvious to me that it should do it.

File: plugins/imagebase/plugin.js [hunk 9]
```diff
@@ -104,6 +274,48 @@
...
+		 *		widgetDefinition = CKEDITOR.plugins.imagebase.addFeature( editor, 'link', widgetDefinition );
+		 *		CKEDITOR.plugins.imagebase.addImageWidget( editor, 'myWidget', widgetDefinition );
+		 *
+		 * @param {CKEDITOR.editor} editor Editor that will get the widget registered.
+		 * @param {String} name Feature name.
+		 * @param {CKEDITOR.plugins.imagebase.imageWidgetDefinition} definition Widget's definition.
+		 * @returns {CKEDITOR.plugins.imagebase.imageWidgetDefinition} Widget's definition extended
+		 * with fields needed by feature.
+		 */
+		addFeature: function( editor, name, definition ) {
```
[the remark this answers is not in the document]
Comment (Comandeer) [reply plugins/imagebase/plugin.js]: It could be a little tricky to do it as our `CKEDITOR.tools.object.merge` always returns new object.

**Comment by mlewand** [comment]
I have rebased the branch, so that we have a cleaner merge. There were some conflicts during that, but I fixed it.
